src/main/java/org/truffleruby/language/methods/LookupMethodOnSelfNode.java

```diff
@@ -31,6 +31,11 @@ public InternalMethod lookup(
         return execute(frame, self, name, config);
     }
 
+    public InternalMethod lookup(
+            Frame frame, Object self, String name, DispatchConfiguration config) {
```

Comment:
Should be `VirtualFrame` here.

Reply:
We already have `VirtualFrame` version: 
https://github.com/oracle/truffleruby/blob/783c3ccf5cd7ff1adbca70185dcdae5cc4c2bb76/src/main/java/org/truffleruby/language/methods/LookupMethodOnSelfNode.java#L29-L37

The problem is that we have `Frame` on the `GetMethodObjectNode` level; So we could type-cast it explicitly or add another `lookup` method as we have there.

Reply:
Can GetMethodObjectNode take `VirtualFrame frame` and `MaterializedFrame callerFrame`? I think that should work, and then we never use Frame which is error-prone (it's a big issue to pass it behind a TruffleBoundary)

Reply:
We discussed this issue and it seems impossible to pass VirtualFrame to `execute` (or `executeGetMethodObject`) method because of `@TruffleBoundary`

```java
    @GeneratedBy(GetMethodObjectNode.class)
    private static final class Uncached extends GetMethodObjectNode {

        private final ContextReference<RubyContext> rubyLanguageContextReference_ = lookupContextReference(RubyLanguage.class);

        @TruffleBoundary
        @Override
        public RubyMethod executeGetMethodObject(VirtualFrame frameValue, Object arg0Value, Object arg1Value, DispatchConfiguration arg2Value, Frame arg3Value) {
            return doGeneric(frameValue, arg0Value, arg1Value, arg2Value, arg3Value, this.rubyLanguageContextReference_.get(), (NameToJavaStringNode.getUncached()), (LookupMethodOnSelfNodeGen.getUncached()), (DispatchNode.getUncached()), (BooleanCastNodeGen.getUncached()), (ConditionProfile.getUncached()), (ConditionProfile.getUncached()), (LogicalClassNode.getUncached()));
        }

error: Method executeGetMethodObject cannot be annotated with @TruffleBoundary and have a parameter of type VirtualFrame
        public RubyMethod executeGetMethodObject(VirtualFrame frameValue, Object arg0Value, Object arg1Value, DispatchConfiguration arg2Value, Frame arg3Value) {
```

So I don't see how to pass VirtualFrame to `LookupMethodOnSelfNode#lookup` without typecasting. We also need to add a public entrypoint to make sure that we have VirtualFrame as frame argument:

```java
public abstract class GetMethodObjectNode extends RubyBaseNode {
   public GetMethodObjectNode getMethodObject(VirtualFrame fame, ...) {
      return execute(frame ...)
   }

   public abstract RubyMethod execute(Frame frame, Object self, Object name,
            DispatchConfiguration dispatchConfig, Frame callerFrame);
}
```

With this, we can typecast frame inside `doGeneric` method. Is this works for you?

Reply:
Mmh, that must be a Truffle DSL bug, it should not add `@TruffleBoundary` for the generated uncached node if there is a VirtualFrame argument.
Yes, let's have a `protected execute(Frame)` then, like in LookupMethodOnSelfNode + that public entrypoint.

Reply:
I discussed this with Truffle people, it seems that typing it as `Frame` is the correct thing to do:
* uncached nodes should never be used in PE code
* so if in PE code, it doesn't matter what kind of frame we pass, it won't escape
* if outside PE code, there is no way for us to have a VirtualFrame instance, so `frame` will be `null` anyway, and it's harmless to call the generated uncached node's execute() (won't make a VirtualFrame escape, which would fail compilation).

Reply:
To clarify a public method taking `Frame` is fine.
I actually have a PR making LookupMethodOnSelfNode#execute public, it should merge in the next days.

Reply:
So you going to replace `LookupMethodOnSelfNode#lookup` via public `LookupMethodOnSelfNode#execute`, right? Ping me when you do this (if it will be before merging this PR)

Reply:
Done in https://github.com/oracle/truffleruby/commit/93211d3fa80390c158a08e9be0f662783e5146ab

Reply:
Could you rebase on top of latest master?

Reply:
Sure; Should I reflect `[GR-28932] Prepare for always-inlined builtins` changes and make `GetMethodObjectNode#execute` public and use it instead of 

```java
public RubyMethod get(VirtualFrame frame, Object self, Object name,
  DispatchConfiguration config, MaterializedFrame callerFrame) {
     return execute(frame, self, name, config, callerFrame); // execute is protected here
}
```
?

Reply:
Yes

Reply:
I'll do that.